Add null safety guidance (#2980)
dart-lang/site-www: src/_guides/language/effective-dart/usage_migrated.md

    expression is `null`. With `== true`, you have to think through the boolean
    logic to realize that means that a `null` gets converted to *false*.

> **Exception:** Using a null-aware operator on a variable inside a condition does
> not promote the variable to a non-nullable type. If you want the variable to be
> promoted inside the body of the if statement, it may be better to use an
> explicit `!= null` check instead of `?.` followed by `??`:

Reviewer: ```suggestion
**Exception:** Using a null-aware operator on a variable inside a condition
doesn't promote the variable to a non-nullable type. If you want the variable
to be promoted inside the body of the `if` statement, it might be better to use
an explicit `!= null` check instead of `?.` followed by `??`:
```

For the may -> might change, see https://developers.google.com/style/word-list#letter-m.

I have to point out that these diffs would've been easier to read if we were using semantic line breaks. (I'll try not to proselytize any more!)

Author: I know semantic line breaks are better. You're absolutely right. They just hurt my soul to look at. I'm sorry.

Reviewer: It helps to think of them as free-form poetry. (I'm probably using that terminology wrong; @RedBrogdon would know the real poetical terms.)

{:.good}
<?code-excerpt "usage_good.dart (null-aware-promote)"?>
{% prettify dart tag=pre+code %}
int measureMessage(String? message) {
  if (message != null && message.isNotEmpty) {
    // message is promoted to String.
    return message.length;
  }

  return 0;
}
{% endprettify %}

{:.bad}
<?code-excerpt "usage_bad.dart (null-aware-promote)"?>
{% prettify dart tag=pre+code %}
int measureMessage(String? message) {
  if (message?.isNotEmpty ?? false) {
    // message is not promoted to String.
    return message!.length;
  }

  return 0;
}
{% endprettify %}

### CONSIDER copying a nullable field to a local variable to enable type promotion.

Checking that a nullable variable is not equal to `null` promotes the variable
to a non-nullable type. That lets you access members on the variable and pass it
to functions expecting a non-nullable type. Unfortunately, promotion is only

> sound for local variables and parameters, so fields and top-level variables are
> not promoted.

Reviewer: ```suggestion
sound for local variables and parameters, so fields and top-level variables
aren't promoted.
```

One pattern to work around this is to copy the field's value to a local

> variable. Null checks on that variable do promote and then you can safely treat

Reviewer: ```suggestion
variable. Null checks on that variable do promote, so you can safely treat
```

it as non-nullable.

{:.good}
<?code-excerpt "usage_good.dart (copy-nullable-field)"?>
{% prettify dart tag=pre+code %}
class UploadException {
  final Response? response;

  UploadException([this.response]);

  @override
  String toString() {
    var response = this.response;
    if (response != null) {
      return "Could not complete upload to ${response.url} "
          "(error code ${response.errorCode}): ${response.reason}.";
    }

    return "Could not upload (no response).";
  }
}
{% endprettify %}

> This can be cleaner and safer than using `!` every place the field or top level
> variable is used.

Reviewer: ```suggestion
Copying to a local variable can be cleaner and safer than using `!` every place
the field or top-level variable is used:
```

{:.bad}
<?code-excerpt "usage_bad.dart (copy-nullable-field)" replace="/!\./[!!!]./g"?>
{% prettify dart tag=pre+code %}
class UploadException {
  final Response? response;

  UploadException([this.response]);

  @override
  String toString() {
    if (response != null) {
      return "Could not complete upload to ${response[!!!].url} "
          "(error code ${response[!!!].errorCode}): ${response[!!!].reason}.";
    }

    return "Could not upload (no response).";
  }
}
{% endprettify %}

> This pattern must be applied with some care. If you need to write back to the
> field make sure you do so, and not just to the local variable. Also, if there
> may be other changes to the field while the local is still in scope, it may end
> up having a stale value. Sometimes it is best to simply use `!` on the field.

Reviewer: ```suggestion
Be careful when using a local copy. If you need to write back to the field,
then make sure you do so, and don't just write to the local variable. Also, if
the field might change while the local is still in scope, then the local might
have a stale value. Sometimes it's best to simply use `!` on the field.
```

    expression is `null`. With `== true`, you have to think through the boolean
    logic to realize that means that a `null` gets converted to *false*.

**Exception:** Using a null-aware operator on a variable inside a condition does
not promote the variable to a non-nullable type. If you want the variable to be
promoted inside the body of the if statement, it may be better to use an
explicit `!= null` check instead of `?.` followed by `??`:

{:.good}
<?code-excerpt "usage_good.dart (null-aware-promote)"?>
{% prettify dart tag=pre+code %}
int measureMessage(String? message) {
  if (message != null && message.isNotEmpty) {
    // message is promoted to String.
    return message.length;
  }

  return 0;
}
{% endprettify %}

{:.bad}
<?code-excerpt "usage_bad.dart (null-aware-promote)"?>
{% prettify dart tag=pre+code %}
int measureMessage(String? message) {
  if (message?.isNotEmpty ?? false) {
    // message is not promoted to String.
    return message!.length;
  }

  return 0;
}
{% endprettify %}

### CONSIDER copying a nullable field to a local variable to enable type promotion.

> Checking that a nullable variable is not `null` promotes it to a non-nullable type so that you can then call methods on the variable. Unfortunately, that is only sound for local variables and parameters, so fields and top-level variables are not promoted.

Reviewer 2: Do we expect readers to understand that "call methods" also covers reading/writing fields/getters?

Author: Reworded.

> Checking that a nullable variable is not `null` promotes it to a non-nullable type so that you can then call methods on the variable. Unfortunately, that is only sound for local variables and parameters, so fields and top-level variables are not promoted.

Reviewer 2: [nit]  "is not `null`" sounds closer to `is! Null` to me than `!= null`. Should this say "Checking that a nullable variable is not equal to `null`"? Or is that too pedantic?

Author: Good point. Reworded.

    expression is `null`. With `== true`, you have to think through the boolean
    logic to realize that means that a `null` gets converted to *false*.

**Exception:** Using a null-aware operator on a variable inside a condition does
not promote the variable to a non-nullable type. If you want the variable to be
promoted inside the body of the if statement, it may be better to use an
explicit `!= null` check instead of `?.` followed by `??`:

{:.good}
<?code-excerpt "usage_good.dart (null-aware-promote)"?>
{% prettify dart tag=pre+code %}
int measureMessage(String? message) {
  if (message != null && message.isNotEmpty) {
    // message is promoted to String.
    return message.length;
  }

  return 0;
}
{% endprettify %}

{:.bad}
<?code-excerpt "usage_bad.dart (null-aware-promote)"?>
{% prettify dart tag=pre+code %}
int measureMessage(String? message) {
  if (message?.isNotEmpty ?? false) {
    // message is not promoted to String.
    return message!.length;
  }

  return 0;
}
{% endprettify %}

### CONSIDER copying a nullable field to a local variable to enable type promotion.

> Checking that a nullable variable is not `null` promotes it to a non-nullable type so that you can then call methods on the variable. Unfortunately, that is only sound for local variables and parameters, so fields and top-level variables are not promoted.

Reviewer 2: [nit] Some other content is formatted to 80 columns. Should this be as well?

Reviewer: Yes, all content except maybe long URLs should be 80 columns or fewer. (We prefer [semantic line breaks](https://github.com/dart-lang/site-shared/blob/master/doc/writing-for-dart-and-flutter-websites.md#semantic-line-breaks), but Bob gets a religious exemption.)

Author: Oops, yes.
